### PR TITLE
Remove IsEquivalent check in RecordSession

### LINKED ISSUE #11632: Re-enable body comparison in StorageRecordMatcher
Test framework now supports request content verification, a few tests might need to be re-recorded to enable it.

cc @seanmcc-msft

### LINKED ISSUE #18903: RecordedTest attribute indicates test was re-recorded when it wasn't
If some of the Event Grid tests fail on playback and needs to be re-recorded, the output indicates that it was re-recorded successfully, but no new session record appears.

### PR DESCRIPTION
**Pull request by christothes**
fixes #18903
closes #11632

### CODE REVIEW

**Comment by pakrym** [comment]
Let's fix the equivalence check to take the body into account instead.

**Comment by JoshLove-msft** [comment]
> Let's fix the equivalence check to take the body into account instead. This seems to be fixing the symptom.

Well that brings up my question from the issue - if the equivalence check is not meant to be more lenient than playback, what purpose does it serve?

**Comment by pakrym** [comment]
It checks if the recording we are going to save is the same as the one already on disk and skips saving it.

**Comment by JoshLove-msft** [comment]
> It checks if the recording we are going to save is the same as the one already on disk and skips saving it.

But why not just use the existing matching logic?

**Comment by pakrym** [comment]
It's too strict, IsEquivalent tries to be a bit more forgiving.

**Comment by christothes** [comment]
I took this approach as a path to totally deprecate the equivalence check. I can look into making it more strict I guess. 

@JoshLove-msft can you give an example test that would cause this same failure? As I recall you thought it may have been not just the body, but the header ordering.

**Comment by JoshLove-msft** [comment]
> I took this approach as a path to totally deprecate the equivalence check. I can look into making it more strict I guess.
> 
> @JoshLove-msft can you give an example test that would cause this same failure? As I recall you thought it may have been not just the body, but the header ordering.

In the Event Grid case it was the ordering of the body contents, so just the body.

**Comment by JoshLove-msft** [comment]
> It's too strict, IsEquivalent tries to be a bit more forgiving.

Well that means that there are cases where we expect record matching to fail, but IsEquivalent to still be true. I'm not sure I understand why we want this behavior. If we made the re-record logic the same as the matching logic, I don't see how this could cause issues as any sessions that would need to be re-recorded would have failed playback.

**Comment by pakrym** [comment]
Imagine the following test:

```
var uploadResult = Upload();
DownloadOnlyIfEtagMatches(uploadResult.ETag);
```

There can exist 2 recordings where both of them would pass playback because ETag is part of the response and is recorded but if compared to one another they won't match because ETags in different recordings are different. IsEquivalent would detect them as being the same.

**Comment by christothes** [comment]
> Imagine the following test:
> 
> ```
> var uploadResult = Upload();
> DownloadOnlyIfEtagMatches(uploadResult.ETag);
> ```
> 
> There can exist 2 recordings where both of them would pass playback because ETag is part of the response and is recorded but if compared to one another they won't match because ETags in different recordings are different. IsEquivalent would detect them as being the same.

It seems that at best, `IsEquivalent` saves us from unnecessarily saving a recording when we are not specific about which tests should be re-recorded. At worst, it causes confusion and inconsistent behavior. I'm wondering if the pros outweigh the cons

**Comment by pakrym** [comment]
>It seems that at best, IsEquivalent saves us from unnecessarily saving a recording when we are not specific about which tests should be re-recorded. 

That was its original goal, yes. 

>At worst, it causes confusion and inconsistent behavior. I'm wondering if the pros outweigh the cons

The default implementation should be strict enough not to break anyone. Services are free to extend it with their service-specific knowledge. I'm OK removing it if we convince Storage to move off it.

**Comment by christothes** [comment]
>  I'm OK removing it if we convince Storage to move off it.

Are you aware of any blocking reason storage couldn't move to using `[RecordedTest]` (assuming someone was motivated to make the change)?

**Comment by pakrym** [comment]
>Are you aware of any blocking reason storage couldn't move to using [RecordedTest] (assuming someone was motivated to make the change)?

I don't know anything particular. If `RecordedTest` fits their workflows they might be OK

**Comment by JoshLove-msft** [comment]
> Imagine the following test:
> 
> ```
> var uploadResult = Upload();
> DownloadOnlyIfEtagMatches(uploadResult.ETag);
> ```
> 
> There can exist 2 recordings where both of them would pass playback because ETag is part of the response and is recorded but if compared to one another they won't match because ETags in different recordings are different. IsEquivalent would detect them as being the same.

Ah I see. It is not just that it is more lenient, it is more lenient because it has to be given that it isn't working off the same data.

File: sdk/core/Azure.Core.TestFramework/src/RecordSession.cs [hunk 1]
```diff
@@ -99,21 +99,6 @@ public void Sanitize(RecordedTestSanitizer sanitizer)
...
-        {
-            if (session == null)
-            {
-                return false;
-            }
-
-            // The DateTimeOffsetNow variable is updated any time it's used so
-            // we only care that both sessions use it or both sessions don't.
-            var now = TestRecording.DateTimeOffsetNowVariableKey;
-            return session.Variables.TryGetValue(now, out string _) == Variables.TryGetValue(now, out string _) &&
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/RecordSession.cs:112]: You also need remove EntryEquivalentComparer and all leftovers from RecordMatcher

Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/RecordSession.cs:112]: And from every library that was overriding them

Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/RecordSession.cs:112]: Deleting code

File: sdk/mixedreality/Azure.MixedReality.Authentication/tests/Azure.MixedReality.Authentication.Tests/MixedRealityRecordMatcher.cs [hunk 2]
```diff
@@ -1,19 +0,0 @@
...
-
-namespace Azure.MixedReality.Authentication.Tests
-{
-    public class MixedRealityRecordMatcher : RecordMatcher
-    {
-        private const string ClientCorrelationVectorHeaderName = "X-MRC-CV";
-
-        public MixedRealityRecordMatcher()
-        {
-            IgnoredHeaders.Add(ClientCorrelationVectorHeaderName);
```
Comment (pakrym) [inline sdk/mixedreality/Azure.MixedReality.Authentication/tests/Azure.MixedReality.Authentication.Tests/MixedRealityRecordMatcher.cs:14]: Can't remove this.

File: sdk/storage/Azure.Storage.Blobs.Batch/samples/Azure.Storage.Blobs.Batch.Samples.Tests.csproj [hunk 3]
```diff
@@ -15,6 +15,9 @@
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
   </ItemGroup>
+  <ItemGroup>
+    <Compile Remove="C:\src\azure-sdk-for-net\sdk\storage\Azure.Storage.Blobs.Batch\samples\..\..\Azure.Storage.Common\tests\Shared\StorageRecordMatcher.cs" />
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs.Batch/samples/Azure.Storage.Blobs.Batch.Samples.Tests.csproj:19]: hm

**Review by pakrym — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs.Batch/samples/Azure.Storage.Blobs.Batch.Samples.Tests.csproj (hunk 3 above)
Comment (christothes) [reply sdk/storage/Azure.Storage.Blobs.Batch/samples/Azure.Storage.Blobs.Batch.Samples.Tests.csproj:19]: StorageRecordMatcher.cs was deleted.

Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs.Batch/samples/Azure.Storage.Blobs.Batch.Samples.Tests.csproj:19]: Sure, so why does this file have a local path in it?

Comment (christothes) [reply sdk/storage/Azure.Storage.Blobs.Batch/samples/Azure.Storage.Blobs.Batch.Samples.Tests.csproj:19]: oh - this is something strange that VS did. I was reading this as being deleted for some reason, but it added this :/

File: sdk/storage/Azure.Storage.Common/tests/Shared/StorageRecordMatcher.cs [hunk 4]
```diff
@@ -1,336 +0,0 @@
...
-            // AD JSON
-            "expires_in",
-            "ext_expires_in"
-        };
-
-        /// <summary>
-        /// Creates a new StorageRecordMatcher to determine whether recordings
-        /// have been meaningfully updated.
-        /// </summary>
-        // TODO: https://github.com/Azure/azure-sdk-for-net/issues/11632
```
Comment (JoshLove-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/StorageRecordMatcher.cs:86]: SHould we close this?

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**